### ISSUE TITLE
Non-nullable type used in examples

### ISSUE THREAD

**Issue by jmewes** [dart-lang/language#2342]
### Observed

The [Dart Language Specification](https://spec.dart.dev/DartLangSpecDraft.pdf) contains the following statement in the section "Classes > Instance Methods > The Method noSuchMethod":

> Note that it must be a method that accepts one positional argument, in order to correctly override noSuchMethod in Object. For instance, it can have signature `noSuchMethod(Invocation i)` or `noSuchMethod(Object i, [String s])`,

However, if a method has the signature `noSuchMethod(Object i, [String s])`, it yields a compiler error:

![compiler-error](https://user-images.githubusercontent.com/5235584/178105310-8ff8d0a1-2db7-4a28-99d8-258851b8fb5b.png)

### Expected

Probably the example is outdated and should use a nullable type for the second parameter.

### References

- https://github.com/dart-lang/language/blame/180b5ba19dc289e78a1341255e17086cd0f90f48/specification/dartLangSpec.tex#L2954

**Comment by srawlins** [comment]
Good catch! I think this spec is for the pre-null safe language though. That text is legal in pre-null safe code.

**Comment by jmewes** [comment]
The "6th edition draft" is for version 2.15-dev which I thought uses null safe by default.

**Comment by eernstg** [comment]
Thanks! Cf. https://github.com/dart-lang/language/pull/2343.

The CHANGELOG comments near the beginning of the language specification LaTeX source are associated with the language version which was the current language version when the change was performed, and that's actually misleading. It's been done like that forever, but we should consider how to choose "section" headers for these change log comments in a way that is more meaningful.

In any case, the plan is that the upcoming stable version of the language specification will specify the language version 2.13. That is, it will include null safety and non-function type aliases. Several PRs are available for that (e.g., #2023, #2052), but it takes time to get them landed. More recent language updates (e.g., super-parameters and enhanced enums) will be integrated into the language specification when the null safety spec has landed.

**Comment by jmewes** [comment]
Also see the first example code snippet in "10.7.1 Generative Constructors" of the Dart Programming Language Specification 6th edition draft:

![Dart_programming_language___Dart](https://user-images.githubusercontent.com/5235584/179420505-3d49afa4-dafc-4690-8245-8a6e226b098d.png)

**Comment by eernstg** [comment]
> Also see

Thanks! This is already handled in the upcoming revision of the language specification where null safety is added, along with many, many other changes.

**Comment by jmewes** [comment]
In regards of the NNBD feature, you might also want to re-consider the following statement in section "18.11 Try":

> An on-catch clause of the form catch (p1, p2) s is equivalent to an on-catch clause on dynamic catch (p1, p2) s.

```dart
void main() {
  try {
    throw 'Foo';
  // ignore: nullable_type_in_catch_clause
  } on dynamic catch (e, t) {
    print(e.runtimeType);
  }
}
```

**Comment by jmewes** [comment]
Here is another snippet from section "15.2 Super-Bounded Types" which needs an update. Maybe it can be check whether it has been updated when the ticket is closed:

![examples_–_scratchpad_dart](https://user-images.githubusercontent.com/5235584/197931993-2ba369ba-4dde-4c6f-aff7-27c9a087bc73.png)

By the way, probably I misunderstood something, but from the wording in the language specification I would have expected this example for super-bounded types to work.

> Types of members from super-bounded class types are computed using the same rules as types of members from other types. Types of function applications involving super-bounded types are computed using the same rules as types of function applications involving other types. Here is an example:

**Comment by eernstg** [comment]
> they are not equivalent anymore

True; the version of the language specification that includes null safety has the following:

```
An on-catch clause of the form `catch (p1 , p2 ) B`
is equivalent to the on-catch clause `on Object catch (p1 , p2 ) B`.
```

So that update has already been made. I've uploaded the complete null-safety PR as https://github.com/dart-lang/language/pull/2605, so please take a look at that one if you wish to check whether any particular rule involving null safety has been updated.

Please note that https://github.com/dart-lang/language/pull/2605 will not be landed as such. It is used to extract smaller PRs (e.g. https://github.com/dart-lang/language/pull/2052). Whenever one of those smaller PRs is landed, the diff between the current version of the language specification and the one in https://github.com/dart-lang/language/pull/2605 will be reduced a bit. When https://github.com/dart-lang/language/pull/2605 is empty, everything has been transferred.

> 15.2 Super-Bounded Types

Good catch! The example needs to initialize `x`, and the super-bounded type must have a top type (`Object` -> `Object?` will do). Those changes are now included in #2605, thanks!

I'll close this issue now, because the normal procedure is to use one issue for one fault or action, and #2605 does already contain the changes needed according to this issue.

@jmewes, thanks a lot for the input! It would be great if you create a new issue if you notice another missing update or other faults!

**Comment by jmewes** [comment]
@eernstg ,

I am glad that those comments are useful and get addressed. I was not sure about that and not sure if I understood it correctly.

For me it is a nice exercise to read the specification. So, I'll create an issue if I find again something that doesn't fit and will try to get it right.
